### DISCUSSION TITLE
Web components in Server components

### DISCUSSION THREAD

**Discussion by stormfar** [vercel/next.js#49537] [App Router]
Prior the /app router and default server components, I was able to use web components generated via StencilJS as per the [with-stencil](https://github.com/vercel/next.js/tree/881d202962b4226330b6fe4986d0cc943e4f40e6/examples/with-stencil), by defining the custom elements in _app.tsx.
```
import { useLayoutEffect } from 'react'
import { applyPolyfills, defineCustomElements } from 'test-component/loader'

export default function App({ Component, pageProps }) {
  useLayoutEffect(() => {
    applyPolyfills().then(() => {
      defineCustomElements(window)
    })
  }, [])
  return <Component {...pageProps} />
}

// Then use in any component like so
export default function Home() {
  return (
    <div>
      <my-component first="Next.js" last="The React Framework"></my-component>
    </div>
  )
}
```

With the advent of default Server components, **is it even possible to use web components (which to my knowledge rely on browser APIs) in server components?**
I tried adapating my project by defining a `CustomElements` client component containing the above `defineCustomElements`, then importing that into a page.tsx server comnponent, but the web components don't render and I get errors like 
```
TypeError: Cannot read properties of undefined (reading '$hostElement$')
TypeError: Cannot read properties of undefined (reading '$instanceValues$')
```
from the web component library.

Any ideas?

**Comment by mbulfair** [comment]
@5tormTrooper I'm looking for something like this for our SSR rendering using the default pages directory. The difference is our group is using the react output binding to import react libraries with stencil components. We have issues currently with warnings of server and client hydration issues. Do you need to do both to hydrate but still be able to import the react wrapped web components? I'm also interested if that would work in react server components.

**Reply by stormfar** [reply]
@mbulfair How are you generating the React bindings? In Stencil or the React project? Are they just warnings are the components not actually interactive?

It looks like Stencil have a [Hydrate App](https://stenciljs.com/docs/hydrate-app) output. May be worth trying out.

**Comment by luwes** [comment]
I don't think much should have changed in this area. The old pages router also SSR's but then tries to hydrate on the client.

With both routers (app and pages) you have to take care of making sure the browser API's used in the custom elements are shimmed or add an if condition (optional chaining can be handy) to prevent type errors when the API's like `globalThis.customElements.define()` are not found.

In our project we add dummy browser shims to prevent these server type errors:
https://github.com/muxinc/media-chrome/blob/main/src/js/utils/server-safe-globals.js

I'm currently working on a project that SSR's web components and its shadow DOM of technically any JS framework including Next.js. 
https://github.com/luwes/wesc

It uses a JSDOM like library to shim most used browser API's so it should work on any server that runs JS.
Please let me know if you try it out or if you run into issues.
